node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/CapacityChecker.java

```diff
@@ -0,0 +1,528 @@
+package com.yahoo.vespa.hosted.provision.maintenance;
+
+import com.yahoo.config.provision.NodeResources;
+import com.yahoo.config.provision.NodeType;
+import com.yahoo.vespa.hosted.provision.Node;
+import com.yahoo.vespa.hosted.provision.NodeRepository;
+import com.yahoo.vespa.hosted.provision.node.Allocation;
+
+import java.util.*;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+public class CapacityChecker {
+    private List<Node> hosts;
+    Map<String, Node> nodeMap;
+    private Map<Node, List<Node>> nodeChildren;
+    private Map<Node, AllocationResources> availableResources;
+
+    public AllocationHistory allocationHistory = null;
+
+    public CapacityChecker(NodeRepository nodeRepository) {
+        this.hosts = getHosts(nodeRepository);
+        List<Node> tenants = getTenants(nodeRepository, hosts);
+        nodeMap = constructHostnameToNodeMap(hosts);
+        this.nodeChildren = constructNodeChildrenMap(tenants, hosts, nodeMap);
+        this.availableResources = constructAvailableResourcesMap(hosts, nodeChildren);
+    }
+
+    public List<Node> getHosts() {
+        return hosts;
+    }
+
+    public Optional<HostFailurePath> worstCaseHostLossLeadingToFailure() {
+        Map<Node, Integer> timesNodeCanBeRemoved = computeMaximalRepeatedRemovals(hosts, nodeChildren, availableResources);
+        return greedyHeuristicFindFailurePath(timesNodeCanBeRemoved, hosts, nodeChildren, availableResources);
+    }
+
+    protected List<Node> findOvercommittedHosts() {
+        return findOvercommittedNodes(availableResources);
+    }
+
+    public List<Node> nodesFromHostnames(List<String> hostnames) {
+        List<Node> nodes = hostnames.stream()
+                .filter(h -> nodeMap.containsKey(h))
+                .map(h -> nodeMap.get(h))
+                .collect(Collectors.toList());
+        if (nodes.size() != hostnames.size()) {
+            List<String> notFoundNodes = hostnames.stream()
```

Comment:
Can be replaced with
```
Set<String> notFoundNodes = new HashSet<>(hostnames);
notFoundNodes.removeAll(nodes);
```

Reply:
No, because `nodes` is a list of `Node`, not `String`.
I can replace it with this version for increased code clarity.
```Java
Set<String> notFoundNodes = new HashSet<>(hostnames);
notFoundNodes.removeAll(nodes.stream().map(Node::hostname).collect(Collectors.toList()));
```